hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs

```diff
@@ -400,23 +411,27 @@ getPlugins = idePlugins <$> getIdeEnv
 -- | 'withProgress' @title f@ wraps a progress reporting session for long running tasks.
 -- f is passed a reporting function that can be used to give updates on the progress
 -- of the task.
-withProgress :: (MonadIde m, MonadIO m) => T.Text -> ((Core.Progress -> m ()) -> m a) -> m a
-withProgress t f = do
+withProgress :: (MonadIde m , MonadIO m, MonadBaseControl IO m)
+             => T.Text -> Core.ProgressCancellable
+             -> ((Core.Progress -> m ()) -> m a) -> m a
```

Comment:
I still think you want the argument to the continuation to be in IO. The other arguments should be generalised though.